rfc/bh-rfc-2.md

```diff
@@ -0,0 +1,114 @@
+-   Nesting gives the AST clean scoping, but that’s for the compiler’s benefit. For humans, deep nesting obscures the happy path and binds unrelated steps together visually. Early returns expose structure more clearly where it matters, at the level of code review, maintenance, and debugging
+
+### 3.2 Considerations for existing patterns
+
+There are valid reasons to use the nesting pattern as it is implemented historically. This section will contain a few justifications and counter points.
+
+#### 3.2.1 Single Exit Rule
+
+Encourage one clear exit point, which can simplify certain types of debugging, logging, or deferred cleanup logic (e.g. closing resources, unlocking mutexes) and can be placed after all the success paths in a single place.
+
+Counterpoint: Go encourages multiple returns for clarity. Single exit rule is a holdover from older (C-like) languages where it mattered. In practice, early returns reduce indentation and make the success path obvious, while cleanup is better handled with defer, which works regardless of exit point.
+
+#### 3.2.2 Inline error handling
+
+Error checks are colocated with the logic they protect, which can help when steps are deeply dependent or when context needs to stay tightly grouped.
+
+Counterpoint: Colocating error checks with logic sounds clean but quickly becomes cluttered, especially as error handling expands (logging, wrapping, metrics). Early returns separate failure paths from the happy path, which aligns with Go’s ‘fail fast’ idiom and improves scannability.
+
+#### 3.2.3 Nested ifs as Scoped Guards:
+
+Each nested block forms a well-scoped zone for a variable, ensuring it can’t accidentally be accessed elsewhere. Helps with tight, explicit control over variable lifetimes.
+Counterpoint: Scoped lifetimes are a weak justification in small, linear functions. We could use anonymous blocks or extract to helper functions. Readability and maintainability matter more than narrowing variable lifespan.
+
+Additionally, this does not guard us completely from scoping issues. If, for example stepC returned only a value (and not an error), the following two lines are now valid.
+
+`} else if c := stepC(); err != nil {`
+`c := stepC() {`
+
+In a stack of nested if statements, this nuance would be easy to miss. The err from stepB is still in scope. While the bug would still compile in the idiomatic flat style, it’s more obvious during review because each variable declaration is isolated and visually prominent
+
+## 4. Summary
+
+Chained nesting offers real scoping guarantees at the compiler level, but it comes at a readability cost for humans. In Go, clarity and simplicity should take precedence. Early returns flatten logic, highlight the success path, and align with the language’s philosophy of small, focused functions and explicit control flow. The goal isn’t just correctness at the machine's level. it’s maintainability at the developer level. And for that, flat wins.
```

Comment:
I'd really like some additional depth in _why_ the idiomatic option is more readable/easier to maintain than nesting. I think there's some valid reasons that can be brought up, but ultimately this is a style discussion, so bringing good points on why something is easier than another is probably the most important piece of the puzzle